### PR TITLE
Make `debug.g/setmetatable` respect the `"__metatable_debug"` metatable field

### PR DESCRIPTION
**Pull request by Desour**
- Goal of the PR: Improvement of the SSM sandbox security.
- How does the PR work:
  * `debug.getmetatable` and `debug.setmetatable` are overwritten in builtin to use the `"__metatable_debug"` field like `getmetatable` and `setmetatable` use the `"__metatable"` field.
  * Userdata metatables have now the same value for `"__metatable_debug"` as for `"__metatable"`. This prevents mods from doing nasty stuff like overwriting or replacing `"__gc"` metatable fields.
    (The `"__metatable"` fields had basically no effect.)
  * (We still have to leave `debug.g/setmetatable` in the sandbox because we want to allow to overwrite the string metatable.)
- Does it resolve any reported issue: Idk.
- If not a bug fix, why is this PR needed? What usecases does it solve? It can be considered a (security-related) bugfix.

## To do

This PR is a Ready for Review.

## How to test

* Check that your mods still work.
* Read the code.
* Read the doc changes.

### CODE REVIEW

**Comment by Desour** [comment]
Now they are not overwritten iff mod security is off. In that case, `_G` is the insecure env, and we don't want to restrict it.

**Comment by rubenwardy** [comment]
I find this change highly questionable. Shouldn't we just remove these functions from the sandbox if there's concern over security?

**Comment by Desour** [comment]
I didn't know how often these functions are used. IIRC there was a discussion in irc about this, where someone (you?) said that they're (almost?) nowhere used. (Ie. mesecons just uses the normal `getmetatable` on strings.) Hence, I'm fine with completely removing them. I kept this PR open as a reminder that the issue still exists. Also IIRC there was someone who wanted the functions to stay to do some weird stuff with functions.

**Comment by sfan5** [comment]
To properly discuss this it'd be useful if someone (@appgurueu?) could sum up on what types `[gs]etmetable` and `debug\.[gs]etmetatable` work.

**Comment by appgurueu** [comment]
> To properly discuss this it'd be useful if someone (@appgurueu?) could sum up on what types `[gs]etmetable` and `debug\.[gs]etmetatable` work.

[`setmetatable`](https://www.lua.org/manual/5.1/manual.html#pdf-setmetatable) only works on tables and respects the `__metatable` field, erroring if called on tables that have it set in their metatable. [`getmetatable`](https://www.lua.org/manual/5.1/manual.html#pdf-getmetatable) also works on other types (strings, for example). It is possible to retrieve metatables set via `debug.setmetatable` using just `getmetatable`. `getmetatable` respects the `__metatable` field and returns its value instead if set, whereas [`debug.getmetatable`](https://www.lua.org/manual/5.1/manual.html#pdf-debug.getmetatable) and [`debug.setmetatable`](https://www.lua.org/manual/5.1/manual.html#pdf-debug.setmetatable) do not respect this field, operating on the raw metatable instead. The `debug.*` functions also work on *pretty much all types* and allow replacing/changing/setting *string, number, boolean, nil* metatables as well as *function* (for all functions). Full userdata, like tables, has individual metatables per instance, which can be set/get using the `debug` metatable funcs.

Example:
```lua
> t = setmetatable({}, {__metatable = 42})
> =getmetatable(t)
42
> =debug.getmetatable(t)
table: 0x55d51c5c4060
> setmetatable(t, {})
stdin:1: cannot change a protected metatable
stack traceback:
	[C]: in function 'setmetatable'
	stdin:1: in main chunk
	[C]: ?
> debug.setmetatable(t, {})
```
Further resources: [pgimeno's metatable tutorial](https://codeberg.org/pgimeno/Gists/src/branch/lua--metatable-tutorial#so-what-s-special-about-a-metatable) and [the Lua reference manual](https://www.lua.org/manual/5.1/manual.html#2.8)

File: builtin/game/secure_sandbox.lua [hunk 1]
```diff
@@ -0,0 +1,34 @@
+
+--[[
+This file contains things for making the global env sandbox secure that weren't
+done in C++.
+]]
+
+-- Overwrite debug.getmetatable and debug.setmetatable, so that they have the same
+-- semantics with the metatable field "__metatable_debug" as getmetatable and
```
Comment (appgurueu) [inline builtin/game/secure_sandbox.lua:8]: Why not respect the `__metatable` field?

File: builtin/game/secure_sandbox.lua [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	local debug_getmetatable_orig = debug.getmetatable
+	local debug_setmetatable_orig = debug.setmetatable
+	-- store error in local variable, so that mods can't overwrite it later
+	local error = error
+
+	debug.getmetatable = function(obj)
+		local mt = debug_getmetatable_orig(obj)
+		if mt.__metatable_debug ~= nil then
+			return mt.__metatable_debug
+		else
```
Comment (appgurueu) [inline builtin/game/secure_sandbox.lua:21]: You don't need the else (early return). You can also turn the `~=` into `==` by swapping the two `return`s. The same applies below.

File: doc/lua_api.txt [hunk 3]
```diff
@@ -23,8 +23,25 @@ functionality in the engine and API, and to document it here.
...
+Also look at the [Lua 5.1 Reference Manual](https://www.lua.org/manual/5.1/).
+
+Changes to the Lua std library
+------------------------------
+
+Due to security reasons, mods are executed in a sandboxed environment (see also
+`minetest.request_insecure_environment()`). Mods therefore do not have access to
+all functions in the std library. To see what is available, iterate over `_G`.
+
+There are also functions that are available, but were changed in their functionality:
```
Comment (appgurueu) [inline doc/lua_api.txt:39]: Either state that this list is non-exhaustive or complete this as [documented here](https://github.com/minetest/minetest_docs/blob/master/doc/environment.adoc)

File: builtin/game/secure_sandbox.lua (hunk 1 above)
Comment (Desour) [reply builtin/game/secure_sandbox.lua:8]: To give more control. And theoretically there could be a mod that does some hacky, but valid, stuff with another mod's object.

If we want to keep `debug.setmetatable` just for non-table objects and nothing else, using `__metatable` might be better.

**Comment by Desour** [comment]
Underlying issue (#12216) was fixed, closing.

**Pull request closed without merging**